From b930a0f1224261068dc7954eddf9163be6dbcdbe Mon Sep 17 00:00:00 2001
Message-Id: <b930a0f1224261068dc7954eddf9163be6dbcdbe.1430046936.git.chang-joon.lee@intel.com>
In-Reply-To: <6f7782ceadb10df792cdc4378455d6e03124c770.1430046936.git.chang-joon.lee@intel.com>
References: <6f7782ceadb10df792cdc4378455d6e03124c770.1430046936.git.chang-joon.lee@intel.com>
From: Deepak S <deepak.s@intel.com>
Date: Thu, 19 Feb 2015 17:42:23 +0530
Subject: [PATCH 2/4] FOR_UPSTREAM [VPG]: drm/i915: Clearing buffer objects
 via blitter engine for Gen8

On Gen8, COLOR BLT commands are different. Add gen8 specific
command to clearing buffer objects via blitter engines.

For: GMINL-7579
Change-Id: I1920a19a649b4ce932336ad5ad8847dbbd5b4eb0
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/gpu/drm/i915/i915_gem_stolen.c |   54 ++++++++++++++++++++++----------
 drivers/gpu/drm/i915/i915_reg.h        |    1 +
 2 files changed, 39 insertions(+), 16 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem_stolen.c b/drivers/gpu/drm/i915/i915_gem_stolen.c
index 03400cb..dfb1b64 100644
--- a/drivers/gpu/drm/i915/i915_gem_stolen.c
+++ b/drivers/gpu/drm/i915/i915_gem_stolen.c
@@ -379,22 +379,44 @@ static int i915_add_clear_obj_cmd(struct drm_i915_gem_object *obj)
 	u32 offset = i915_gem_obj_ggtt_offset(obj);
 	int ret;
 
-	ret = intel_ring_begin(ring, 6);
-	if (ret)
-		return ret;
-
-	intel_ring_emit(ring, COLOR_BLT_CMD |
-			      XY_SRC_COPY_BLT_WRITE_ALPHA |
-			      XY_SRC_COPY_BLT_WRITE_RGB);
-	intel_ring_emit(ring, BLT_DEPTH_32 | (PAT_ROP_GXCOPY << ROP_SHIFT) |
-			PITCH_SIZE);
-	intel_ring_emit(ring,
-			(DIV_ROUND_UP(obj->base.size, PITCH_SIZE) <<
-						 HEIGHT_SHIFT) | PITCH_SIZE);
-	intel_ring_emit(ring, offset);
-	intel_ring_emit(ring, 0);
-	intel_ring_emit(ring, MI_NOOP);
-	intel_ring_advance(ring);
+	if (IS_GEN8(dev_priv->dev)) {
+		ret = intel_ring_begin(ring, 8);
+		if (ret)
+			return ret;
+
+		intel_ring_emit(ring, GEN8_COLOR_BLT_CMD |
+				      XY_SRC_COPY_BLT_WRITE_ALPHA |
+				      XY_SRC_COPY_BLT_WRITE_RGB | (7-2));
+		intel_ring_emit(ring, BLT_DEPTH_32 |
+				      (PAT_ROP_GXCOPY << ROP_SHIFT) |
+				       PITCH_SIZE);
+		intel_ring_emit(ring, 0);
+		intel_ring_emit(ring, obj->base.size >> PAGE_SHIFT <<
+						HEIGHT_SHIFT | PAGE_SIZE / 4);
+		intel_ring_emit(ring, i915_gem_obj_ggtt_offset(obj));
+		intel_ring_emit(ring, 0);
+		intel_ring_emit(ring, 0);
+		intel_ring_emit(ring, MI_NOOP);
+
+		intel_ring_advance(ring);
+	} else {
+		ret = intel_ring_begin(ring, 6);
+		if (ret)
+			return ret;
+
+		intel_ring_emit(ring, COLOR_BLT_CMD |
+				      XY_SRC_COPY_BLT_WRITE_ALPHA |
+				      XY_SRC_COPY_BLT_WRITE_RGB);
+		intel_ring_emit(ring, BLT_DEPTH_32 | (PAT_ROP_GXCOPY <<
+					ROP_SHIFT) | PITCH_SIZE);
+		intel_ring_emit(ring,
+				(DIV_ROUND_UP(obj->base.size, PITCH_SIZE) <<
+						HEIGHT_SHIFT) | PITCH_SIZE);
+		intel_ring_emit(ring, offset);
+		intel_ring_emit(ring, 0);
+		intel_ring_emit(ring, MI_NOOP);
+		intel_ring_advance(ring);
+	}
 
 	return 0;
 }
diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 28d9315..79beb83 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -378,6 +378,7 @@
 #define XY_SRC_COPY_BLT_CMD		((2<<29)|(0x53<<22)|6)
 #define XY_MONO_SRC_COPY_IMM_BLT	((2<<29)|(0x71<<22)|5)
 #define COLOR_BLT_CMD			((2 << 29) | (0x40 << 22) | 3)
+#define GEN8_COLOR_BLT_CMD		(2<<29 | 0x50<<22)
 #define XY_SRC_COPY_BLT_WRITE_ALPHA	(1<<21)
 #define XY_SRC_COPY_BLT_WRITE_RGB	(1<<20)
 #define   BLT_DEPTH_8			(0<<24)
-- 
1.7.9.5

